File: leetcode/139-word-break.js
// https://leetcode.com/problems/word-break/ (top-interview-questions)

/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function(s, wordDict) {
  
  const dict = new Set(wordDict);

  const memo = {};
  memo[s.length] = true;

  const partitioning = (i) => {

    if (memo.hasOwnProperty(i)) return memo[i];

    for (let j = i+1; j <= s.length; j += 1) {
      const subStr = s.slice(i, j);
      if (!dict.has(subStr)) {
        continue;
      }

      if(partitioning(j)) {
        memo[i] = true;
        return true;
      }
    }

    memo[i] = false;
    return false;

  }
  
  return partitioning(0);
};

File: leetcode/139-word-break.py
# https://leetcode.com/problems/word-break/ (top-interview-questions)

from typing import List

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        tabulation = [False] * (len(s)+1)
        tabulation[len(s)] = True

        for i in reversed(range(len(s))):
            for word in wordDict:
                if i+len(word) < len(tabulation) and s[i: i+len(word)] == word and tabulation[i+len(word)] == True:
                    tabulation[i] = True
                    break
                    
        print(tabulation)
        return tabulation[0]
